src/cpp/subscriber/StatisticsParticipantListener.cpp

```diff
@@ -238,6 +238,91 @@ StatisticsParticipantListener::StatisticsParticipantListener(
 {
 }
 
+// Return true if the address is localhost: 127.0.0.1
+bool is_local_host(
+        octet* address)
+{
+    return (address[12] == 127 && address[13] == 0 && address[14] == 0 && address[15] == 1);
+}
+
+// Convert an address to an IP format: "37.11.18.30"
+std::string address_to_string(
+        octet* address)
+{
+    return std::to_string((int)address[12]) + "." + std::to_string((int)address[13]) + "." +
+           std::to_string((int)address[14]) + "." + std::to_string((int)address[15]);
+}
+
+// Search for an address different from localhost in the locator list
+bool search_address_in_locators(
+        const eprosima::fastrtps::ResourceLimitedVector<Locator_t>& locators,
+        std::string& address)
+{
+    for (auto locator : locators)
+    {
+        if (!is_local_host(locator.get_address()))
+        {
+            address = address_to_string(locator.get_address());
+            return true;
+        }
+    }
+    return false;
+}
+
+// Return a IP obtained from participant locators
+std::string get_address(
+        const ParticipantProxyData& info)
+{
+    // The IP is obtained from the announced locators
+    // Search for a locator with an IP different from localhost
+    std::string address;
+
+    // 1. default_locators.unicast
+    if (search_address_in_locators(info.default_locators.unicast, address))
+    {
+        return address;
+    }
+
+    // 2. metatraffic_locators.unicast
+    if (search_address_in_locators(info.metatraffic_locators.unicast, address))
+    {
+        return address;
+    }
+
+    // 3. default_locators.multicast
+    if (search_address_in_locators(info.default_locators.multicast, address))
+    {
+        return address;
+    }
+
+    // 4. metatraffic_locators.multicast
+    if (search_address_in_locators(info.metatraffic_locators.multicast, address))
+    {
+        return address;
+    }
+
+    // The only option is for localhost to be the only valid IP
+    return "127.0.0.1";
+}
+
+// Return the participant_id
+std::string get_participant_id(
+        const GUID_t& guid)
+{
+    // The participant_id can be obtained from the last 4 octets in the GUID prefix
+    std::stringstream buffer;
+    for (int i = 0; i < 3; i++)
+    {
+        buffer << std::hex << std::setfill('0');
+        buffer << std::setw(2) << static_cast<unsigned>(guid.guidPrefix.value[i + 8]);
+        buffer << ".";
+    }
+    buffer << std::hex << std::setfill('0');
```

Comment:
```suggestion
    buffer << std::hex << std::setfill('0');
    for (int i = 0; i < 3; i++)
    {
        buffer << std::setw(2) << static_cast<unsigned>(guid.guidPrefix.value[i + 8]);
        buffer << ".";
    }
```

It is redundant to set again the format of the buffer. Unless you set the flags with other values, the settings persist.